### PR TITLE
minimega: add time in state to vm info and web interface

### LINKED ISSUE #811: minimega: uptime per each VM
Create a per-VM uptime record.  May exist but not shown on web.

### PR DESCRIPTION
**Pull request by jasoncontrib**
Closes #811.

### CODE REVIEW

File: src/minimega/vm.go [hunk 1]
```diff
@@ -620,8 +635,10 @@ func (vm *BaseVM) Info(field string) (string, error) {
 // setState updates the vm state, and write the state to file. Assumes that the
 // caller has locked the vm.
 func (vm *BaseVM) setState(s VMState) {
-	log.Debug("updating vm %v state: %v -> %v", vm.ID, vm.State, s)
+	now := time.Now()
+	log.Debug("updating vm %v state: %v -> %v, at time %s", vm.ID, vm.State, s, now)
```
Comment (jcrussell) [inline src/minimega/vm.go]: `log` includes the current time -- it doesn't seem necessary to add it to the message.

File: src/minimega/vm.go [hunk 2]
```diff
@@ -568,6 +581,8 @@ func (vm *BaseVM) Info(field string) (string, error) {
 		return vm.Namespace, nil
 	case "state":
 		return vm.State.String(), nil
+	case "statetime":
```
Comment (jcrussell) [inline src/minimega/vm.go]: Maybe `updated` makes more sense that `statetime` and instead of `time.Since`, report the actual time that the VM was mutated?

File: src/minimega/vm.go [hunk 3]
```diff
@@ -44,6 +45,7 @@ type VM interface {
 	GetNetworks() []NetConfig // GetNetworks returns an ordered, deep copy of the NetConfigs associated with the vm.
 	GetHost() string          // GetHost returns the hostname that the VM is running on
 	GetState() VMState
+	GetStateStart() time.Time // GetStateStart returns the time when the VM entered the current state
```
Comment (jcrussell) [inline src/minimega/vm.go]: Glad `miniweb` will allow us to drop a lot of these accessor methods.

File: src/minimega/vm.go (hunk 2 above)
Comment (jasoncontrib) [inline src/minimega/vm.go]: Used `time.Since` since @bjwrigh asked for something like `uptime`, which reports a duration?

**Comment by djfritz** [comment]
I read the discussion in #811 as just the total wallclock uptime from `vm launch`. What's the use case to reset the time at each state change?

**Comment by jasoncontrib** [comment]
I read @bjwrigh 's original request for `uptime` as wanting to know wallclock time that a VM had been `RUNNING`, and figured we might as well show it for other states than `RUNNING`. Perhaps we can get some clarification from @bjwrigh ?

**Comment by bjwrigh** [comment]
Only the RUNNING state seems necessary.  Clearing the time when the VM leaves the RUNNING state would probably help keep things accurate.  I don't see an added benefit to seeing how long something was BUILDING, other than maybe using that as a trigger to do a distributed start on the VM's for resource purposes - but that information can be discerned from the level of the script running the experiment.

tl;dr:  My opinion is to only track RUNNING time, reset to n/a or zero when not in that state.

**Comment by jasoncontrib** [comment]
Would you want to see time as duration elapsed, or the date/time at which it started in the RUNNING state?

________________________________

**Comment by bjwrigh** [comment]
I don't know that I have a way of choosing which one would be best in any way other than preference.  The elapsed time may be deceiving on a stale info page, while the start time might be deceiving if no date is present, or the system clock on one node of the cluster is off.  Both would lead to it being too cluttered.  I'm leaning towards uptime... @djfritz thoughts?

**Comment by jcrussell** [comment]
It seems like elapsed time makes the most sense if we can't count on system clocks being correct/synchronized.

@jasoncontrib: I would tweak this to be called `uptime` and only apply to VMs in the running state.

However, if a VM is paused/started, should the uptime revert to 0?

**Comment by djfritz** [comment]
I still don't see the point. If I pause and later unpause a QEMU vm, the vm will see the jump in time. If the uptime is just time in the running state, what possible use is there?

My use case for something like this is to have a simple timestamp of when the vm was launched, rendered in vm info as an elapsed wall clock time from that launch point. I'd use that as an indicator of things like when I launch a new experiment and find some old VM lingering and messing up my experiment.

File: src/minimega/vm.go [hunk 4]
```diff
@@ -302,6 +307,14 @@ func (vm *BaseVM) GetState() VMState {
 	return vm.State
 }
 
+func (vm *BaseVM) GetStateStart() time.Time {
+	vm.lock.Lock()
+	defer vm.lock.Unlock()
+
+	log.Info("Returning StateStart: ", vm.StateStart.String())
```
Comment (jcrussell) [inline src/minimega/vm.go]: Don't think this log is necessary.

**Comment by jcrussell** [comment]
@djfritz: If that's the case, then we should store `Started` and always compute `time.Since(vm.Started)`?

File: src/minimega/vm.go [hunk 5]
```diff
@@ -44,6 +45,7 @@ type VM interface {
 	GetNetworks() []NetConfig // GetNetworks returns an ordered, deep copy of the NetConfigs associated with the vm.
 	GetHost() string          // GetHost returns the hostname that the VM is running on
 	GetState() VMState
+	GetLaunchTime() time.Time // GetLaunchTime returns the time when the VM entered the current state
```
Comment (jcrussell) [inline src/minimega/vm.go]: Forgot to update comment.
